cdap-securestore-ext-cloudkms/src/main/java/co/cask/cdap/securestore/gcp/cloudkms/CloudKMSClient.java

```diff
@@ -0,0 +1,248 @@
+        .setKeyRingId(KEYRING_ID)
+        .execute();
+    } catch (GoogleJsonResponseException e) {
+      // if key ring already exists, then do not throw any exception.
+      if (e.getDetails() != null && e.getDetails().getCode() == 409) {
+        LOG.trace(String.format("Key ring %s already exists", KEYRING_ID));
+        return;
+      }
+      throw new IOException(String.format("Exception occurred while creating key ring %s", KEYRING_ID), e);
+    }
+  }
+
+  /**
+   * Creates a new crypto key on google cloud kms with the given id.
+   *
+   * @param cryptoKeyId crypto key id
+   * @throws IOException if there's an error creating crypto key
+   */
+  void createCryptoKeyIfNotExists(String cryptoKeyId) throws IOException {
+    // If crypto key is already created, do not attempt to create it again.
+    if (knownCryptoKeys.contains(cryptoKeyId)) {
+      return;
+    }
+
+    String parent = String.format("projects/%s/locations/%s/keyRings/%s", projectId, LOCATION_ID, KEYRING_ID);
+
+    CryptoKey cryptoKey = new CryptoKey();
+    // This will allow the API access to the key for symmetric encryption and decryption.
+    cryptoKey.setPurpose(ENCRYPT_DECRYPT);
+
+    try {
+      cloudKMS.projects().locations().keyRings().cryptoKeys()
+        .create(parent, cryptoKey)
+        .setCryptoKeyId(cryptoKeyId)
+        .execute();
+    } catch (GoogleJsonResponseException e) {
+      // Crypto key is shared for all the secrets in a namespace. If the crypto key already exists, then do not throw
+      // any exception. This will happen if another key for the same namespace is being created.
+      if (e.getDetails() != null && e.getDetails().getCode() == 409) {
+        LOG.trace(String.format("Key %s already exists", cryptoKeyId));
+        return;
+      }
+
+      throw new IOException("Error occurred while creating cryptographic key for namespace %s" , e);
+    }
+
+    // In-memory cache to keep list of crypto keys created so far.
+    knownCryptoKeys.add(cryptoKeyId);
+  }
+
+  /**
+   * Encrypts secret with provided crypto key.
+   *
+   * @param cryptoKeyId crypto key to encrypt secret
+   * @param secret secret to be encrypted
+   * @throws IOException there's an error in encrypting secret
+   */
+  byte[] encrypt(String cryptoKeyId, byte[] secret) throws IOException {
+    String resourceName = String.format("projects/%s/locations/%s/keyRings/%s/cryptoKeys/%s", projectId, LOCATION_ID,
+                                        KEYRING_ID, cryptoKeyId);
+    // secret must not be longer than 64KiB.
+    EncryptRequest request = new EncryptRequest().encodePlaintext(secret);
+    EncryptResponse response = cloudKMS.projects().locations().keyRings().cryptoKeys()
+      .encrypt(resourceName, request)
+      .execute();
+
+    byte[] encryptedData = response.decodeCiphertext();
+    if (encryptedData == null) {
+      throw new IOException("Error while encrypting the secret. Encrypted data is null.");
```

Comment:
Does the `EncryptResponse` contains any error message about why the data is null? If it does, worth to include that in the message.

Reply:
No it does not have any error message about data being null

Reply:
that's disappointing..